design/ceph/multus-network.md

```diff
@@ -78,7 +78,123 @@ Nothing to do in particular since they don't use any service IPs.
+
+In order to not leave moved interfaces dangling in the host's network namespace, mover pods must
+move interfaces back to their original namespace when CSI is being terminated. The most
+straightforward way to accomplish this is to move interfaces back when the mover is being
+terminated. If an interface is moved without user applications also being removed, this will cause
+I/O disruption. Therefore, the DaemonSet should also use the `OnDelete` update strategy so that the
+pods can be updated on node reboots (or node drains).
+
+In order to better handle unexpected corner cases that leave moved interfaces in the host network
+namespace (e.g., a mover is killed abruptly rather than gracefully terminated), instead treat "move"
+operations as a disable-and-copy operation. To do this, disable the interface in the holder pod's
+network namespace, and create a copy of the interface in the host namespace with the same MAC
+address and IP config. From the user standpoint, the interface is still "moved" because the original
+is disabled, so we keep the "mover" terminology. This merely helps Rook ensure that it is not
+accidentally losing the original information.
+
+A previous iteration of this design specified the mover application as a sidecar to Ceph CSI plugin
+pods; however, this design would mean that the mover would need to be deleted and re-created
+whenever the CSI plugin is updated, possibly resulting in I/O hangs in user pods during the update.
+Keeping the mover independent allows CSI plugin updates to happen freely without complex
+interactions between it and the mover.
+
+#### Interactions between components
+If a copied interface is left in the host network namespace after the holder pod is removed, multus
+may later give the address to a different application, and the CSI driver may try to connect to the
+different application with Ceph requests. We should try to avoid leaving interfaces on the host as
+much as possible. Killing the mover pod abruptly will leave copied interfaces, but there is no way
+to prevent this from happening.
+
+If a holder pod is deleted, the interface hold will be lost. The mover must remove the
+interface from the host's network namespace because multus may reassign the address to a different
+application. We can document for users that they should not delete holder pods, but we cannot
+prevent users from manually stopping holder pods. However, to prevent the Kubernetes scheduler from
+terminating holder pods, the pods should be given the highest possible priority so that they are not
+un-scheduled except by users.
+
+There is a possible race condition where a mover pod is killed and where a holder is deleted before
+a new mover starts up. In this case, the copied interface for the holder pod will be left in the
+host network namespace, but the mover will not get a notification that the holder was removed. In
+order to clean up from this case, upon startup, the mover should delete all interface copies in the
```

Comment:
Hum, this looks risky, what if another app is doing the exact same thing as we do?

Reply:
I think this is a risk for any operation the "mover" is doing. If another application (either in a container or on the host) is modifying network interfaces, it could interfere with our mover.